### PR TITLE
Use a select with options for the books in tests_run

### PR DESCRIPTION
**Pull request by dav1312**
Easier for the users to select the book that they want without having to manually check the repo and type it correctly

image

### CODE REVIEW

**Comment by vondele** [comment]
hmm, since most users should not be changing the book, I'm not sure that listing the full choice is a good idea, it will lead to more random book testing. While it is possible to test different books, we don't want to encourage it, I think. So, not 100% sure.

**Comment by dav1312** [comment]
> it will lead to more random book testing
I think most of the fishtest contributors are "responsible" in that regard.
Tests still need to pass through approvers and they can raise their concerns and delete tests if a user starts requesting all sorts of random books for no real reason.

Currently there is not really a way to know which books are available for testing. You have to figure out that the books are the ones in the stockfish org by luck or by looking at the source code.

**Comment by peregrineshahin** [comment]
you can just make it available for admin or approver group with your pending PR https://github.com/glinscott/fishtest/pull/1383

**Comment by vdbergh** [comment]
_if_ it has to be restricted then I would favor an "Advanced" tick box enabling it (like the tick box for time odds matches).

File: server/fishtest/views.py [hunk 1]
```diff
@@ -490,6 +490,19 @@ def get_master_bench():
     return None
 
 
+def get_valid_books():
+    response = requests.get(
+        "https://api.github.com/repos/official-stockfish/books/contents"
+    ).json()
+    return list(
```
Comment (ppigazzini) [inline server/fishtest/views.py:497]: In modern python an iterable comprehension (or in this case a memory saving generator expression) is replacing the use of functional programming elements such as `map`, `filter`, see
https://stackabuse.com/functional-programming-in-python/

`endswith()` accepts a tuple of strings to be searched for
https://docs.python.org/3/library/stdtypes.html?highlight=endswith#str.endswith

```python
    books_list = (
        b["path"].replace(".zip", "")
        for b in response
        if b["path"].endswith((".epd.zip", ".pgn.zip"))
    )
    return books_list
```

DEV running with this change.

**Review by ppigazzini — APPROVED** [review]

**Comment by ppigazzini** [comment]
PROD updated, thank you @dav1312 :)

**Comment by vondele** [comment]
I like that the drop down is behind a tick box.. nice.

**Comment by vdbergh** [comment]
Something that might be considered is to shorten the list by giving only one name, without extension, per book. Whether a book is epd or pgn is an implementation detail that should not concern the user.

**Comment by dav1312** [comment]
> Something that might be considered is to shorten the list by giving only one name, without extension, per book. Whether a book is epd or pgn is an implementation detail that should not concern the user.

Well, the thing is that pgn books can have custom depths that the user can customize that their epd counterpart doesn't have

**Comment by vdbergh** [comment]
> > Something that might be considered is to shorten the list by giving only one name, without extension, per book. Whether a book is epd or pgn is an implementation detail that should not concern the user.
> 
> Well, the thing is that pgn books can have custom depths that the user can customize that their epd counterpart doesn't have

Ok I hadn’t thought of this. So dropping the extension is not a good idea. Still it seems .pgn is strictly better than .epd. So one might say that there is no point showing the .epd file if a .pgn file exists.

**Pull request merged**